### ISSUE TITLE
closed source?

### ISSUE THREAD

**Issue by frostworx** [sipeed/NanoKVM#1]
interesting device, found it via https://www.cnx-software.com/2024/07/08/20-nanokvm-is-a-tiny-low-power-risc-v-kvm-over-ip-solution/#comments

but are you really planning to keep it closed source?

edit: finally ordered one 2 month later :)

**Comment by frostworx** [comment]
a bit random grepping reveals that at least

https://github.com/hodgef/react-simple-keyboard
is used in the firmware

`server/web/assets/index-u2n1Wr7j.css: *   https://github.com/hodgef/react-simple-keyboard`

**Comment by frostworx** [comment]
see also
https://news.ycombinator.com/item?id=40903461

**Comment by Zepan** [comment]
We are wondering wheter open or not.
At the early time (3month ago), we want make it opensource, only make the hardware, and port PiKVM on it with few develop work. but we realize it is not easy to run PiKVM on it, as it not support V4L2.
We send the hardware sample to some community developer who familiar with PiKVM, and they worked for 3month to port V4L2, and still working...(may finished in few month, when it finished, we will open PiKVM port)
we have to develop FW ourself, it takes about 2~3month, and we see >95% commit of PiKVM is written by author, so we think it may not help much if we opensource at the first time? keep this as a bug report repo maybe better?
Welcome feed back if you have better suggestion to work with community, we want a good balance with opensource and commerce.

**Comment by Zepan** [comment]
At least, we will open most interfaces/API for images,HID operations, that DIYer can customize their own functions without dig hard into complex bsp sdk.

**Comment by Zepan** [comment]
I make a vote on twitter: https://x.com/SipeedIO/status/1810508366336115192

**Comment by frostworx** [comment]
thank you for the comprehensive answer and also thanks for creating the poll.
_(personally I don't use x-twitter, so I won't take part)_

**Comment by ThomasKaiser** [comment]
Not using X/Twitter either but wanting to emphasis on the open source aspect -> trust into the device being not equipped with a backdoor.

Folks are fine using whatever commercial iKVM solution but once it's open hardware but comes from **China** they start to worry. Especially residents of the [Five Eyes](https://en.wikipedia.org/wiki/Five_Eyes) states who are absolutely fine with their governments/agencies spying on everyone on this planet are trained to believe that everything that originates from China must be evil. 

As such allowing users to build their own firmware image from sources will at least be a try to stop these 'concerns' your KVM device being backdoored by design.

**Comment by shangri26199** [comment]
+1 for opensource without X account

**Comment by seichold** [comment]
I bought a full and lite through the pre-order to check them out, but without the firmware being open source I do not think this will gain wide traction. 

I really want this project to be successful as it brings a near order of magnitude cost reduction (and size!) to the ip based KVM world . 

So take this as my (non X using) vote to work toward open source firmware.

**Comment by ElbPirat** [comment]
Also +1 on open source.
Pre-Ordered one for personal use, but searching something similar for my workshop/lab at work, other professional devices are too expensive for my superior... Would try to order a few once released.

**Comment by bilogic** [comment]
i need HDMI to a monitor as well, is it possible with this hardware? or splitter?

**Comment by denern** [comment]
I have a LicheeRV Nano board and am familiar with PiKVM, how can I help you to finish?

**Comment by M0NsTeRRR** [comment]
I'm waiting for the code to be open source before buying one :)

**Comment by joubin** [comment]
> evil.



> > Not using X/Twitter either but wanting to emphasis on the open source aspect -> trust into the device being not equipped with a backdoor.
> > Folks are fine using whatever commercial iKVM solution but once it's open hardware but comes from **China** they start to worry. Especially residents of the [Five Eyes](https://en.wikipedia.org/wiki/Five_Eyes) states who are absolutely fine with their governments/agencies spying on everyone on this planet are trained to believe that everything that originates from China must be evil.
> > As such allowing users to build their own firmware image from sources will at least be a try to stop these 'concerns' your KVM device being backdoored by design.
> 
> what about open backend(go), close frontend(react)?


I know many downvoted this. But if you’re willing to open source the backend and sell an SDK for enterprises that want to build their own front end anyways and integrate with their own Idps. Not sure. There might be something there

**Comment by brainstorm** [comment]
Here's another strategy that can be a win-win for both Sipeed and regular consumers of your product:

**Commit to releasing it as opensource as soon as you've sold, say, 3000 units**. Motivation behind this is:

1. It "Secures" a (fixed, minimum) income threshold for Sipeed.
2. It motivates those buyers to purchase your product to accelerate it being opensource faster.

As it stands right now I'm siding with @M0NsTeRRR's position: if there is no clear commitment to release it as opensource, I'm not interested since I know that **sooner or later** the binary will be reverse engineered and re-implemented as OSS anyway (plus all the other security and future maintenance related concerns outlined in this issue).

**Comment by timonoj** [comment]
I bought it first, assuming it was OSS from the beginning. I'm gonna wait for the OSS release. I can see them adding new features to their private fw every few days, but I'd rather very much just use OSS.

**Comment by scpcom** [comment]
I already took some look at LicheeRV-Nano-Build and this is what I found so far:
- freertos: contains pre-built cvitek audio libraries, did not take a closer look if this is required
- host-tools: pre-built compilers (no problem for me, can be replaced by distro/standard compliers)
- linux_5.10, opensbi and u-boot-2021.10: fully open source, can be built outside of LicheeRV-Nano-Build with minimal modifications
- middleware: contains pre-built libraries, I could not find the source for it
- osdrv: contains two pre-built modules (soph_vc_driver.ko and soph_jpeg.ko) but the source code was already released by cvitek/sophgo
- ramdisk: contains pre-built versions of standard packages like busybox (no problem for me, can be replaced by distro/standard versions)

The middleware should also be open-sourced (by the SoC vendor), it seems to be required for NanoKVM

**Comment by scpcom** [comment]
I almost finished my research on LicheeRV-Nano-Build and created the branches on forks of existing projects:

https://github.com/scpcom/linux/tree/licheervnano-5.10.y
https://github.com/scpcom/opensbi/tree/licheervnano-0.9
https://github.com/scpcom/u-boot/tree/licheervnano-2021.10

https://github.com/scpcom/buildroot/tree/licheervnano-2023.11
https://github.com/scpcom/FreeRTOS/tree/licheervnano

https://github.com/scpcom/sophgo-fsbl/tree/licheervnano
https://github.com/scpcom/sophgo-ramdisk/tree/licheervnano

https://github.com/scpcom/sophgo-build/tree/licheervnano
https://github.com/scpcom/sophgo-middleware/tree/licheervnano
https://github.com/scpcom/sophgo-osdrv/tree/licheervnano

Currently I am locally connecting them together as submodules, I may create and publish a top-level repo next, to make it easier.

If you look in the history you can see a replay of the commits to LicheeRV-Nano-Build splitted in to the sub projects.
They are similar with one exception:
The most interesting part is the 2.6G sized initial commit named "release" (bd39c05f39ab6d850d7c2c29a6b5ac617efc698a).
If you look in the history of my branches the size is reduced to less than 2M. You can easily watch it on github or in an editor.
It contains only the changes relevant for LicheeRV Nano now.

First I needed to find a point on the original project which is close to the current LicheeRV Nano version.
Then I made a diff and searched for more parts to exclude. I put things like wireless drivers into separate commits.
The goal was to get a better history and find the hardware relevant changes, the resulting code is identical.

Some notes about the content and references to the original projects:

buildroot
sophgo does also have a 2023.11 branch, but it s not the same:
https://github.com/sophgo/buildroot-2021.05/tree/buildroot-2023.11
It is clean 2023.11.2 plus overlay and some extra packages (downloads wifi firmware blobs and pre-built ld-musl-riscv64 libs)
https://github.com/buildroot/buildroot

freertos
Based on sophgo version with five extra tasks: camera, isp, vcodec, vi and vip
https://github.com/sophgo/freertos (https://github.com/FreeRTOS/FreeRTOS)
https://github.com/sophgo/FreeRTOS-Kernel (https://github.com/FreeRTOS/FreeRTOS-Kernel)
https://github.com/FreeRTOS/Lab-Project-FreeRTOS-POSIX
The binaries are just temp files from the compiler.

fsbl
No modifications, just one extra directory symlink (plat/sg200x)
https://github.com/sophgo/fsbl

host-tools
Not part of LicheeRV-Nano-Build, just:
https://github.com/sophgo/host-tools
compilers and toolchain

linux_5.10
sophgo linux_5.10 is very close to the XUANTIE-RV version but they decided to start
there branch with a blank 5.10.4 and mix XUANTIE-RV and there changes into one big commit
https://github.com/sophgo/linux_5.10/ (https://github.com/XUANTIE-RV/linux)

opensbi
Only few lines changed
https://github.com/XUANTIE-RV/opensbi

oss
Identical to:
https://github.com/sophgo/oss
Does not seem to contain sources, just older pre-compiled libs with headers and pkgconfig.

ramdisk
Contains some extra static linked binaries and configuration.
Two extra firmware blobs in rootfs/common_arm/usr/share/fw_vcodec,
I do not know if the blobs are really relevant or just leftovers from older sdk version.
https://github.com/sophgo/ramdisk

u-boot-2021.10
https://github.com/sophgo/u-boot-2021.10

The delta between LicheeRV-Nano-Build and the following repos (also some parts of the others)
is a bit dirty because sipeed used some sdk with a different naming scheme for the chips.
On the sophgo git versions they use "cv181x" in almost all cases for the chip used in LicheeRV Nano.
On the sdk used by sipeed they reference 3 different names in the code and directory/file names: cv181x, mars and sg200x.

build
https://github.com/sophgo/build
middleware
https://github.com/sophgo/middleware
osdrv
https://github.com/sophgo/osdrv

middleware
Still the repo with most blobs:
middleware/v2/modules/audio/lib
middleware/v2/modules/isp/lib
middleware/v2/3rdparty/cli

middleware/v2/sample/test_mmf
Contains a pre-compiled lib named libmaix_mmf.a (only for riscv64)
If you look in the history of LicheeRV-Nano-Build, you may find a file named
sophgo_middleware.c which seems to be/was the source code for it (and sophgo_middleware.h is equal to maix_mmf.h).

middleware/v2/sample/test_mmf/media_server-1.0.1
Contains fragments based on:
https://github.com/ireader/media-server
https://github.com/ireader/avcodec
https://github.com/ireader/sdk

If you add the missing parts you may be able to compile the libs found in
middleware/v2/sample/test_mmf/media_server-1.0.1/libs

**Comment by brainstorm** [comment]
Good reversing writeup on:

https://lichtlos.weblog.lol/2024/08/how-to-reverse-the-sipeed-nanokvm-firmware

**Comment by scpcom** [comment]
Here is my version of the Nano-Build repo using the submodules described above:
https://github.com/scpcom/LicheeSG-Nano-Build

**Comment by polyzium** [comment]
The frontend was recently open sourced, and the backend will be open sourced when the repository reaches 2k stars or the NanoKVM sells 10k units. [See this tweet](https://x.com/SipeedIO/status/1821891038266482911)

**Comment by M0NsTeRRR** [comment]
> The frontend was recently open sourced, and the backend will be open sourced when the repository reaches 2k stars or the NanoKVM sells 10k units. [See this tweet](https://x.com/SipeedIO/status/1821891038266482911)

A good start, glad to see that !

**Comment by ckotte** [comment]
They should open source everything in the first hand and not wait for whatever. I bought one too because I'm interested, but I can understand why people not want to buy it if it's running on closed source.

**Comment by MatthewCroughan** [comment]
How am I supposed to trust a device that gives backdoor access to my hardware, without the ability to run or maintain my own operating system or software for it?

**Comment by polyzium** [comment]
> How am I supposed to trust a device that gives backdoor access to my hardware, without the ability to run or maintain my own operating system or software for it?

The NanoKVM is based on the LicheeRV Nano. If you want to make your own firmware, feel free, the information is located [here](https://wiki.sipeed.com/hardware/en/lichee/RV_Nano/1_intro.html).
That being said, I saw your [tweet](https://xcancel.com/SipeedIO/status/1823371168822296945#m) regarding concerns about stolen PiKVM code. The answer is no, their firmware is original and does not use PiKVM code. In fact, they even tried to port PiKVM to the NanoKVM hardware (if they are to be believed), but I assume it is either unoptimized or the hardware is a bit weak for it.
I also suggest you look at [this article](https://lichtlos.weblog.lol/2024/08/how-to-reverse-the-sipeed-nanokvm-firmware) regarding reverse engineering of the firmware and software inside. I am pretty sure that the firmware is original again, considering that most of PiKVM is written in Python+C, while Sipeed's implementation uses Golang.
As for the security concerns, Sipeed has [acknowledged them](https://www.hackster.io/news/security-researcher-warns-on-sipeed-s-nanokvm-finds-vulnerabilities-and-a-cat-in-the-firmware-e1157a9ff0f4) and will release a stable and more secure firmware in the near future. The product listing says "beta version" for a reason.

**Comment by MatthewCroughan** [comment]
@polyzium I saw those claims that they tried to port kvmd and related code, which is why I am surprised to see that they have none in their end product. Overall the release process for this thing hasn't been transparent enough for what the product does (provide backdoor access to your hardware). I'm aware of all of the things I can do myself, but I haven't got my hands on the hardware yet, despite applying for the beta boards. I was looking forward to putting NixOS and kvmd on the device, or building open firmware with Nix, for it.

**Comment by polyzium** [comment]
> which is why I am surprised to see that they have none in their end product

IIRC they said it would be an option in the future. The reason it's not out yet is because it's too slow to be usable.

> Overall the release process for this thing hasn't been transparent enough for what the product does (provide backdoor access to your hardware).

I'm not sure what you mean by "backdoor". Do you mean an actual backdoor set up by the manufacturer so that the Chinese can spy on you, or the very function that this device does (basically VNC, but at the hardware level)? In the case of the former, I'd advise analyzing the traffic and blocking any calling home IPs with something like a PiHole.

> I was looking forward to putting NixOS and kvmd on the device, or building open firmware with Nix, for it.

What is the point? Nix is a package manager (not to mention with a STEEP learning curve) and the firmware/rootFS doesn't even have a package manager to begin with. Also, this device is many times weaker than a Raspberry Pi, and the firmware consists of a barebones Buildroot rootFS with Busybox. It's like putting Ubuntu on a microcontroller. Let the device do what it's designed to do.

**Comment by MatthewCroughan** [comment]
> I'm not sure what you mean by "backdoor"

An IP KVM is a way of getting into hardware you control out-of-band. It is a backdoor into your network and your hardware. One that only you should control.

> What is the point? Nix is a package manager 

Nix is something that lets you create a reproducible operating system and deployment that you control, that is easy to maintain, keep up to date, rollback in the event of mistakes. Its methodology makes porting software and targeting new hardware much easier. I think the advantages of using it are obvious.

> Also, this device is many times weaker than a Raspberry Pi, and the firmware consists of a barebones Buildroot rootFS with Busybox

See the following:

- https://www.liminix.org
- github.com/cleverca22/not-os

> It's like putting Ubuntu on a microcontroller

No, it isn't.

**Comment by polyzium** [comment]
> An IP KVM is a way of getting into hardware you control out-of-band. It is a backdoor into your network and your hardware. One that only you should control.

I wouldn't put it that way. I would say that an IPKVM is a hardware-level remote desktop solution, like I said, basically hardware-level VNC. Most rack-mount servers like HPE ProLiant have IPKVM built in, so following your logic, are you saying that every rack-mount server has a backdoor? That would sound VERY alarming. Sipeed is clearly marketing this device as an IPKVM designed to "access your computer from anywhere". So I would consider the whole "not transparent enough" argument invalid.

> Nix is something that lets you create a reproducible operating system and deployments that you control, that is easy to maintain, keep up to date, rollback in the event of mistakes. Its methodology makes porting software and targeting new hardware much easier. I think the advantages of using it are obvious.
> See the following:

Whatever you say, this is a device you own and you are free to do whatever you want with it.
I can't judge NixOS because of my limited experience with it. The only thing I can say is that I tried NixOS in a VM and 2 hours later my brain was melting; common conventions like FHS don't even literally exist in NixOS, they are thrown out the window, and that's where the steep learning curve starts. But I digress.
If I wanted a reproducible system, I'd use a container or pre-built images.

**Comment by MatthewCroughan** [comment]
> are you saying that every rack-mount server has a backdoor

Yes, that only the owner of that system controls. They're also called out-of-band management solutions, but I prefer to call them a backdoor, to keep myself conscious that they are a security critical piece of equipment that if mismanaged would compromise the rest of the system, since whoever has access to the backdoor has access at a lower level.

> Sipeed is clearly marketing this device as an IPKVM that allows you to "access your computer from anywhere"

"Access your device from anywhere".

Using their closed source firmware, that isn't reproducible or modifiable by the consumer.  I think it's understandable that I difficulty trusting this, because you have to trust something, and this is hard to trust for me.

>  If I wanted a reproducible system, I'd use a container or pre-built images.

For me, it cannot be ignored that container images are more popularly used to distribute artifacts of unreproducible build processes.

A preference for pre-built images rather than sources is usually an indicator that something lacks reproducibility. Because otherwise you'd just provide sources and a reproducible build processes, which you can of course then use to create an image for use with containers.

Sometimes images are provided for convenience, or as part of a deployment that requires bit-for-bit determinism to be applied, like A/B updates to fleets of devices, and have legitimate uses. This is not what I'm referring to. Images can be outputs of reproducible build processes, but often images are provided instead of reproducible build processes.

Storing, distributing, and relying solely upon images of unreproducible build processes, rather than providing sources and reproducible build processes, means that making modifications to the original software, or even validating that the image is derived from original software sources, is more difficult, reducing the overall auditability of the software stack.

The TL;DR of that, is that just because you can tar something up and give it to me, doesn't mean I can re-apply the process that created all the things you tarred up, given only the sources. That is much harder, and is the domain of reproducible builds, and things like Nix. Calling images "reproducible" is a misnomer, I think.

Given all of the above, you can see why I might want to address that and build a reproducible operating system for this KVM, as I already have for the Pi CM4 based devices. I don't have access to the hardware, though I may have the ability to play with one soon, since someone who pre-ordered it might have it by the time we meet up for an event.

**Comment by sammcj** [comment]
@sipeed do you have any updates on the full source for this? It seems like you're shipping half baked products at the moment.

**Comment by hadleyrich** [comment]
> @sipeed do you have any updates on the full source for this? It seems like you're shipping half baked products at the moment.

https://github.com/sipeed/NanoKVM/issues/1#issuecomment-2278831374

**Comment by scpcom** [comment]
> > are you saying that every rack-mount server has a backdoor
> 
> Yes, that only the owner of that system controls. They're also called out-of-band management solutions, but I prefer to call them a backdoor, to keep myself conscious that they are a security critical piece of equipment that if mismanaged would compromise the rest of the system, since whoever has access to the backdoor has access at a lower level.

Only the owner of that system? Does anyone have source code for Dell iDRAC or HPE iLO? Can you install your own firmware?

> > Sipeed is clearly marketing this device as an IPKVM that allows you to "access your computer from anywhere"
> 
> "Access your device from anywhere".
> 
> Using their closed source firmware, that isn't reproducible or modifiable by the consumer. I think it's understandable that I have difficulty trusting this, because you have to trust something, and this is hard to trust for me.

The LicheeRV Nano firmware is open and reproducable (as you can read above). And you can customize it.
Currently there are only two components that are not fully open: middleware and the NanoKVM web server. And this is what we are talking about here.

**Comment by MatthewCroughan** [comment]
@scpcom 

> Only the owner of that system? Does anyone have source code for Dell iDRAC or HPE iLO?

Precisely. Scary, isn't it? :)

> The LicheeRV Nano firmware is open and reproducable (as you can read above)

Have you got a link to a git repository containing the source code you're talking about, just so we're clear that I know what you're talking about.

WebServer and Middleware are still two pieces of closed source software that are running as root according to lichtlos' reverse engineering blog post. The webserver is a huge 13MB binary, so it could be quite literally doing anything at all, no question about that.

Would be nice to confirm they are running as root with `ps aux`, though I don't have a device to test anything yet.

**Comment by scpcom** [comment]
> > Only the owner of that system? Does anyone have source code for Dell iDRAC or HPE iLO?
> 
> Precisely. Scary, isn't it? :)

Are you joking or can you give me a link to the sources?

> Have you got a link to a git repository containing the source code you're talking about, just so we're clear that I know what you're talking about.

This is the main repository:
https://github.com/sipeed/LicheeRV-Nano-Build
Just scroll up for full description:
https://github.com/sipeed/NanoKVM/issues/1#issuecomment-2270716285

**Comment by MatthewCroughan** [comment]
> Are you joking or can you give me a link to the sources?

No, I'm agreeing with you that it's scary that the sources aren't available to such critical equipment?

**Comment by scpcom** [comment]
OK, I agree this is scary. To be fair, Dell seems to release some of the sources:
https://opensource.dell.com/releases/idrac9/

But the chance to build and run your own firmware is much better on NanoKVM, I hope to see the missing parts here in the next weeks :-)

**Comment by geerlingguy** [comment]
I just received a couple of these devices to review (I have been in contact with Sipeed over the past couple months)—I am very excited about the hardware, but like others in this thread share concerns over recommending it (or even running it outside a locked down vlan) because _even if there are no geopolitical backdoor concerns_, I like to easily see the provenance of the code I run on all my critical systems.

I also heard the PiKVM code may even run on this hardware, which is great, even if slower—it would at least allow an alternative if Sipeed doesn't end up open sourcing all the code.

Some people aren't as worried about open vs. closed source, but I think especially with NanoKVM being a RISC-V-adjacent project, it would be much easier to recommend with everything being as open as possible.

**Comment by polyzium** [comment]
Hi @geerlingguy! Did not expect to see you here! Love your channel!

The hardware is based on the LicheeRV Nano running on a RISC-V SoC, and there's plenty of documentation on it, so if you want to write your own firmware, you can. The fact that Sipeed wrote their own firmware from scratch does raise such concerns, but considering that it's slowly becoming open source--not to mention a PiKVM port--it gives some hope for the future of this project. Hell, the RISC-V architecture itself is open, so it would make sense that the entire ecosystem of hardware, software and firmware working in symbiosis would be open.

**Comment by scpcom** [comment]
> I also heard the PiKVM code may even run on this hardware, which is great, even if slower—it would at least allow an alternative if Sipeed doesn't end up open sourcing all the code.
> 
> Some people aren't as worried about open vs. closed source, but I think especially with NanoKVM being a RISC-V-adjacent project, it would be much easier to recommend with everything being as open as possible.

I think with PiKVM or any other alternative we still need the libs to encode the video stream with hardware acceleration.

Currently I am wiping all not required closed source libs from LicheeRV-Nano-Build. I also found some additional sources for required libs that can be integrated.

But 7 closed source libs are still required (for the KVM Demo that I use to test it):

middleware/v2/modules/isp
libae
libaf
libawb
libisp_algo
libjson-c (blob also found in json-c.tar.gz)

oss/oss_release_tarball
libcvi_json-c (from cvi-json-c.tar.gz)
libcvi_miniz (from cvi-miniz.tar.gz)

I will provide more information (and git commits) when I am done testing it on the real hardware.

**Comment by CRCinAU** [comment]
I've got a pre-order in for 5 of the full units. Would be good to see the code opened so I can build my own if/as required.

**Comment by scpcom** [comment]
The source code for  soph_jpeg.ko and soph_vc_driver.ko is tested and working now.
I also added the missing components for NanoKVM to build your own image:
https://github.com/scpcom/LicheeRV-Nano-Build/tree/develop
Related pull requests:
https://github.com/sipeed/LicheeRV-Nano-Build/pull/44
https://github.com/sipeed/LicheeRV-Nano-Build/pull/45

During this work I found another lib that is required for nanokvm and not part of the zip file:
/kvmapp/kvm_system/dl_lib/libmaixcam_lib.so
If the file is missing it gets downloaded if you login to the web interface.
I could not find a download link for this special version.
The source code is part of MaixCDK. I added the pre-built version from MaixCDK to the buildroot until I know how to compile it. This is only useful for first boot to get the OLED display of NANOKVM full working.
But the HDMI stream works only after you pressed the update button in the web interface (like on the original NanoKVM image).

**Comment by scpcom** [comment]
OK, libmaixcam_lib seems to be an enhanced C++ version of libmaix_mmf.
But I was wrong: The source code is NOT part of MaixCDK, there is an option that says "compile maixcam_lib from souce code" but only a pre-built version and headers is available.
Is there a plan to release the source code too?

**Comment by petestanley** [comment]
I'm new here and trying to keep pace with the rapid developments. This seems like hot info that alludes to more definitive intentions around the product OSS:

**Source**: https://wiki.sipeed.com/hardware/en/kvm/NanoKVM/system/introduction.html#Open-Source-Repository
_Supposed page time stamp of today (2024-09-04)_.

> NanoKVM frontend is now open source, and the backend will be open-sourced soon (after the GitHub repository reaches 2K stars).

>PiKVM Support Coming Soon

^ Unclear how much SiPeed is contributing vs PiKVM community.

CC: @IronicBadger @geerlingguy

**Comment by ironicbadger** [comment]
Lots of eyes on this project atm given the price point. The outcomes from this [security researcher](https://www.hackster.io/news/security-researcher-warns-on-sipeed-s-nanokvm-finds-vulnerabilities-and-a-cat-in-the-firmware-e1157a9ff0f4) will be much easier to stomach once the code is open.

Hopefully these commitments get followed through on. So far sipeed has given us no reason to think they won't be. 🚀

**Comment by captainwasabi** [comment]
Hasn't anyone run this on a network with Wireguard to see if it's phoning home?  or where it's phoning to (probably AWS then on to whereever).

**Comment by pdostal** [comment]
Wireshark?

**Comment by captainwasabi** [comment]
lol, yeah fixed.  I keep doing that

**Comment by seang96** [comment]
[According to this](https://github.com/sipeed/NanoKVM/issues/54#issuecomment-2357462981) source code is expected to be released by the time you'd get your shipped, so its probably not really worth analyzing its packets at this point.

**Comment by Tea23** [comment]
> The frontend was recently open sourced, and the backend will be open sourced when the repository reaches 2k stars or the NanoKVM sells 10k units. [See this tweet](https://x.com/SipeedIO/status/1821891038266482911)

Almost at 2k. Are you sticking to this?

**Comment by geerlingguy** [comment]
I'm seeing 2K stars now!

<img width="213" alt="Screenshot 2024-09-22 at 8 43 13 PM" src="https://github.com/user-attachments/assets/4c811711-ec40-401d-83e6-27be17b9345d">

**Comment by gshpychka** [comment]
> I'm seeing 2K stars now!
> 
> <img alt="Screenshot 2024-09-22 at 8 43 13 PM" width="213" src="https://private-user-images.githubusercontent.com/481677/369733344-4c811711-ec40-401d-83e6-27be17b9345d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcwNzYzODksIm5iZiI6MTcyNzA3NjA4OSwicGF0aCI6Ii80ODE2NzcvMzY5NzMzMzQ0LTRjODExNzExLWVjNDAtNDAxZC04M2U2LTI3YmUxN2I5MzQ1ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkyM1QwNzIxMjlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00MTBkNzg3NTMzZTBkZDZiYTY0NDg5ZDkyOTFkODBjMWUyZTJhNWJmNWYxMmZmM2UyZTJiM2JhYTRhNTcyZjJiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9._6sqNrIzzwSX_4i_kDs8u0Z7ELiNhpOKLrQ6l_zCVWo">

Almost :)

<img width="254" alt="image" src="https://github.com/user-attachments/assets/b3cd0ff1-74bc-4ea0-b9ed-61012e28b771">
https://github.com/sipeed/NanoKVM/stargazers

**Comment by anscarlett** [comment]
I'd be tempted to remove my star once the source is released...

**Comment by mikeeq** [comment]
Seems like SipeedIO will open-source the backend mid-October :tada: 

"[@SipeedIO](https://x.com/SipeedIO)
[#RISCV](https://x.com/hashtag/RISCV?src=hashtag_click) [#NanoKVM](https://x.com/hashtag/NanoKVM?src=hashtag_click) github repo star is coming to 2K! We will organize and opensource backend source code in Mid October！(and it is also the time all preoreders ship out)"

https://x.com/SipeedIO/status/1838059618720948452

**Comment by onionhammer** [comment]
Good job @geerlingguy :D

**Comment by J-Siu** [comment]
Waiting for update/announcement😁

**Comment by gshpychka** [comment]
> Waiting for update/announcement😁

There has already been one - check above. (Targeting mid-October)

**Comment by scpcom** [comment]
I updated my submodule based repo and did larger changes to get the firmware to a more recent level:
https://github.com/scpcom/LicheeSG-Nano-Build/tree/develop

- switched to sophgo weekly rls 2024.09.11
- switched to buildroot to 2024.05.3
- merged mainline kernel v5.10.226
- added media_server submodule and rtsp_server source from MaixCDK to middleware/sample/test_mmf

Some more details can be found in the README.md

In addition to the 7 libs listed here...
https://github.com/sipeed/NanoKVM/issues/1#issuecomment-2299189750
...the source code for libcvi_bin_isp is included but I kept using the pre-built version because it must match with the older (but optimized) isp_tuning binaries found in /mnt/cfg/param/ and I was not able yet to make the open source version compatible.
nanokvm and other apps maybe linked to more closed binaries but the code is not used - especially audio libs and libcli (telnet server?) can be removed from Makefiles.
For a full open source solution we may need source code for 7 cvi/sophgo libs and sipeed libmaixcam_lib (almost 80% to 90% maybe already available as maix_mmf in middleware/sample/test_mmf) and upcoming nanokvm app source :-)

I updated the "/mnt/system/usr/bin/test_mmf 10" sample to get a stable rtsp stream (stable but still with some delay), no OLED or other display required.

Tested on LicheeRV Nano W with gc4653 and NanoKVM.

**Comment by askmrsinh** [comment]
Looks like the server code just [went up on GitHub](https://github.com/sipeed/NanoKVM/tree/3b2e142657d24337d8b3cedefdaea0777d81ec0d/server)!

**Comment by scpcom** [comment]
Added to buildroot:
https://github.com/scpcom/buildroot/tree/nanokvm-2024.05/package/nanokvm-server
kvm_stream and kvm_system ist not published yet? :-)

**Comment by mtlynch** [comment]
I saw this release on Twitter:

https://x.com/SipeedIO/status/1844035902596956574

And the linked folder has this note:

>10. With regard to the open source note
>
>As the current basic engineering is based on the company openbmc project based on the transplant, so the current release of the mirror of the source project will not be open-source, the follow-up time may be based on the openbmc community original project to pull a version to merge into the open-source again, but only part of the open-source (related to the company's related technology can not be open-source out of the content [such as kvm basedon the vnc of the h265 encoding and decoding etc.], will only be open source in the form of a binary bin file).

This is a violation of the OpenBMC project (mixture of MIT and GPL licenses) and likely any other open-source code this project relies on.

If you're compiling or including open-source code into your app, obeying the license is not optional, and you can't distribute the license "in the form of a binary bin file."

Read the terms of the license of the projects you're including. MIT and GPL both require license notices for each project you're including/compiling, and GPL requires you to provide sources, including any modifications or derivations.

**Comment by geerlingguy** [comment]
@mtlynch - This particular issue is regarding the software they're currently running on the NanoKVM — the linked Twitter post is about an experimental OpenBMC port, that I don't believe is being supported or recommended at this point, just mentioned.

It would probably be best to open a new issue to discuss that port.

The source code for NanoKVM's backend is now in this repository—see https://github.com/sipeed/NanoKVM/tree/3b2e142657d24337d8b3cedefdaea0777d81ec0d/server (however also some things might not be fully integrated, see @scpcom's comment above).

**Comment by gshpychka** [comment]
I expect to be able to flash the device completely from scratch with everything built from source - that's the only way NanoKVM could be truly considered open source. Hopefully we'll get that mid-october as promised.

**Comment by shaver** [comment]
Until we see concrete evidence otherwise, I think we should assume a good-faith effort on Sipeed's behalf to get their system opened up. From what I've seen they are doing real, difficult, uncompensated work to satisfy those of us who prefer open source for network control components, while still meeting their previous commitments to the customers who bought the devices already, and keeping the business afloat.

Sipeed has been unusually transparent about how the source license interacts with their business needs, and the decisions that they faced around that licensing. I think it was tremendously generous of them to offer open sourcing based on source demand (GitHub stars) and not just business milestones like recovery of R&D costs or similar. I feel that this has earned them some patience and grace from the community, and I hope that they see the comments here making demands of them as being excessive enthusiasm rather than hostility. I don't want them to regret their decision to offer the open source path.

**Comment by 174n** [comment]
Any plans on open-sourcing the firmware itself or will it stay closed-source forever?

**Comment by J-Siu** [comment]
@174n The source tree was updated a few days ago, and README is also updated. However, I am not knowledgeable enough to verify it has everything to compile. Maybe @scpcom can confirm?

**Comment by scpcom** [comment]
Yes, you can compile NanoKVM server and web and it works on the hardware.
But the sources for 3 binaries are still missing:
- kvm_system: Responsible for OLED output and some few copy operations (prepares the kvm_stream folder).
- kvm_stream: Runs a MJPEG stream listening on 127.0.0.1:8000
- libmaixcam_lib: Abstraction layer used by kvm_stream to get HDMI input and encode it to jpeg.

Since I also have a plain Lichee RV Nano and MaixCDK did not work I started to create a replacement for libmaixcam_lib based on the found sources some weeks ago, now it is a fully functional replacement (only vdec missing, not required for NanoKVM):
https://github.com/scpcom/sophgo-middleware/tree/maix_mmf-cvisdk/sample/test_mmf/maixcam_lib
https://github.com/scpcom/sophgo-middleware/tree/maix_mmf-cvisdk/sample/test_mmf/maix_mmf

Next step was not much to do, just using the code from test_mmf sample to get a kvm_system replacement:
https://github.com/scpcom/sophgo-middleware/tree/maix_mmf-cvisdk/sample/kvm_system

Some days ago I wanted to know what was happening on port 8000 and at the end I found a cool project that works as perfect replacement:
https://github.com/scpcom/streameye/tree/kvm_stream

I would prefer the original code but it is optional now :-)

**Comment by 174n** [comment]
> you can compile NanoKVM server and web

I meant the firmware. The stuff that gets flashed to the device from the .img file.
I heard from someone that they were going to release the firmware source code sometime in October, so I was wondering if they would release the firmware after they released server's source code

**Comment by shaver** [comment]
Perhaps this is a naive question, but isn't the server flashed to the device as part of the image?

**Comment by scpcom** [comment]
> > you can compile NanoKVM server and web
> 
> I meant the firmware. The stuff that gets flashed to the device from the .img file. I heard from someone that they were going to release the firmware source code sometime in October, so I was wondering if they would release the firmware after they released server's source code

The current official firmware image is created with this:
https://github.com/sipeed/LicheeRV-Nano-Build
The base firmware image was already open source since beginning. But NanoKVM does not work without the NanoKVM app.
The server (app) in the official image is included but does (did) not work without pressing the update button in the web interface to download a dongled lib.
If you are thinking about PiKVM or OpenBMC: This is not officially released or supported yet. There is only a test image of openbmc.

**Comment by burner-** [comment]
> If you are thinking about PiKVM or OpenBMC: This is not officially released or supported yet. There is only a test image of openbmc.

Do you meant that install openBMC to nanoKVM hardware or nanokvm firmware to some of openbmc hardware?

**Comment by J-Siu** [comment]
@scpcom Thank you for the info! As your repo can serve as a fall back for future maintenance, I am ordering now😁

**Comment by scpcom** [comment]
Some conclusion about my work done.

If you want to use a fork of the original project:
https://github.com/scpcom/LicheeRV-Nano-Build/tree/develop
- Easier to fork, less updated components
- Builds NanoKVM server, web, maixcam_lib, kvm_system and kvm_stream from source

If you want to use a more recent base system:
https://github.com/scpcom/LicheeSG-Nano-Build/tree/develop
- Complex git with submodules, much updated components
- Builds NanoKVM server, web, maixcam_lib, kvm_system and kvm_stream from source
- Updated buildroot, linux and sophgo sdk

Building an SD/TF card image is similar for both:
- Run on Ubuntu 22.04 or higher is recommended
- Clone the repostirory
- Execute:
`./build-nanokvm.sh`

During the build process the original latest.zip is downloaded but no binaries are used from it anymore. It is just needed for the init scripts and config files.

The resulting image works out of the box, NanoKVM itself does not need any internet connection to work, no additional downloads required.

**Comment by scpcom** [comment]
The make the build more reproducible I removed the download of latest.zip and added the non-binary files to a git instead:
https://github.com/scpcom/buildroot/tree/nanokvm-2024.05/package/nanokvm-sg200x
https://github.com/scpcom/nanokvm-skeleton

I also updated the build script for NanoKVM server and web:
https://github.com/scpcom/buildroot/tree/nanokvm-2024.05/package/nanokvm-server

On NanoKVM 2.1.0 the kvm_stream binary was replaced with libkvm.so binary:
https://github.com/sipeed/NanoKVM/tree/main/server/dl_lib

I started my own implementation here:
https://github.com/scpcom/sophgo-middleware/tree/maix_mmf-cvisdk/sample/test_mmf/kvm_vision

Most magic ist still done inside maixcam_lib:
https://github.com/scpcom/sophgo-middleware/tree/maix_mmf-cvisdk/sample/test_mmf/maix_mmf
https://github.com/scpcom/sophgo-middleware/tree/maix_mmf-cvisdk/sample/test_mmf/maixcam_lib

And the middleware libs are also still required.

The h264 implementation in my maixcam_lib variant is complete (including avc2flv converter).
My h264 implementation in kvm_vison is just a draft at this moment
(m)jpeg is working fast and stable.

**Comment by scpcom** [comment]
h264 is working with my kvm_vision variant too now.

**Comment by AkechiShiro** [comment]
@scpcom Is it possible that your changes make it upstream to the official firmware/linux, basically into this repository?

**Comment by VFansss** [comment]
I know this will sound quite annoying, but I'm not able to truly understand all the comments and what happened with trying to fully open source the firmware/libraries: is someone buy a NanoKVM today, what kind of things can use/flash onto it to possibly only use open source components? There's still something closed source that must be used? What's the degree of risk using it, eventually?

Sorry for this "non technical" question but as everyone else I'm concerned about security of the device, and I'm not able to truly understand (as someone that doesn't have it right now, but plan to buy it if everything matches) the degree of uncertainty/obscurity if I use the device like it is vs what's available as totally open source

**Comment by scpcom** [comment]
> @scpcom Is it possible that your changes make it upstream to the official firmware/linux (maybe not all changes but the improvements at least), basically into this repository?

I already created two pull requests in September:
https://github.com/sipeed/LicheeRV-Nano-Build/pull/44
https://github.com/sipeed/LicheeRV-Nano-Build/pull/45
Since there was no response I keep them opened and up-to-date but did not try to open new pull requests.

> if someone buy a NanoKVM today, what kind of things can use/flash onto it to possibly only use open source components? There's still something closed source that must be used? What's the degree of risk using it, eventually?

If you use the original sources from Sipeed you have to put them all together yourself, and not all official sources are published yet, you still would require some closed source libs.
This is the reason why I created my own fork, put it all together and created open source replacements for the few missing libs. If you follow this comment you are able build your own image from source:
https://github.com/sipeed/NanoKVM/issues/1#issuecomment-2443169104

**Comment by scpcom** [comment]
Let's talk about the 7 closed source libs remaining.
This is not limited to Sipeed, they are part of cvitek/sophgo middleware.

libae.so, libaf.so, libawb.so: responsible for things like gain control
libisp_algo.so: algos for color correction, filters and profiles
libcvi_json-c.a, libcvi_miniz.a and libjson-c.so: used to parse isp_tuning sensor parameters found in /mnt/cfg/param/

If you have a Lichee RV Nano with camera module (a.k.a. MaixCAM) you may want to keep them.
But do we need them for NanoKVM? No.

Many of the lib symbols are referenced but during runtime of NanoKVM only 3 of them are really used in addition to the (Un)Register functions:
CVI_ISP_GetExposureAttr
CVI_ISP_SetExposureAttr
CVI_ISP_SetWDRExposureAttr

I did not want to harm the source code which is referencing ae, af and awb so I created dummy libs that almost do nothing:
https://github.com/scpcom/sophgo-middleware/tree/maix_mmf-cvisdk/modules/dummy

algo, json-c and miniz is only used in libcvi_bin.so, libcvi_bin_isp.so and libisp.so.
There are existing complier falgs that allow to stop usig them by just adding a Makefile:
https://github.com/scpcom/sophgo-middleware/tree/maix_mmf-cvisdk/modules/isp_light

On NanoKVM you will see not difference in quality, speed or functionality.
I made sure the dummy and light libs are only used inside the kvmapp folder, all other projects still can use the fully featured versions.
At this stage NanoKVM runs full open source now.

Something about licenses: I put my own commits under BSD 3-clause.
But since all work is based on existing code or part of existing project you have to check the source tree for the license used there.

**Comment by scpcom** [comment]
All other pre-built host-tools like genimage are also build from source now, provided by linux, u-boot and buildroot:
https://github.com/scpcom/buildroot/commit/02116db9743d1e532ede3cb0c5fa0f7a817bfcb0
https://github.com/scpcom/sophgo-build/commit/682a4db05e3cf744eb6a0d1f83584412e6d2e917
https://github.com/scpcom/sophgo-build/commit/cec9ba15d752f2d53828ca4ff8b789d6b1336d65
https://github.com/scpcom/sophgo-build/commit/f19053c46f07c81156a00f041decb8992cad9b1e
Now you do not need a AMD/Intel x86_64 host anymore to build the image. I verified the build is working at least on ARM aarch64 machines too.

Building tailscale from source is now also possible with LicheeSG-Nano-Build:
https://github.com/scpcom/buildroot/blob/nanokvm-2024.05/package/tailscale-riscv64/tailscale-riscv64.mk

You can enable it with:
`./build-nanokvm.sh --tailscale`
I keep it disabled by default because tailscaled is permanently calling the tailscale servers even if you not configured it and there is no option yet in the NanoKVM web interface to completely disable it.

**Comment by bt90** [comment]
It's rather sad to compare this botched attempt at open sourcing piece by piece with competitors like JetKVM, which manages to provide the full source code in a single repo.

https://github.com/jetkvm/kvm

**Comment by Frogomeli** [comment]
> It's rather sad to compare this botched attempt at open sourcing piece by piece with competitors like JetKVM, which manages to provide the full source code in a single repo.
> 
> https://github.com/jetkvm/kvm

It's out of scope from this issue, but i don't see the source code of `jetkvm_native` used in `native.go`. Maybe I'm missing something.

**Comment by scpcom** [comment]
As far as I can see the jetkvm repo is just the app (web server) which is similar to the repository here: Can not be used standalone.
There is no bootloader (opensbi, u-boot etc), no kernel and no operating system. It maybe a alternative (if the hardware is available some day) but can not compared with the project(s) discussed here: providing a full firmware image for an existing product.

**Comment by geerlingguy** [comment]
From their [Kickstarter update](https://www.kickstarter.com/projects/jetkvm/jetkvm/posts/4282996):

> The only missing piece is the system image repo, which will be added in the next few days, as we’re finalizing some essential details before a public release.

**Comment by bt90** [comment]
> but can not compared with the project(s) discussed here: providing a full firmware image for an existing product.

And that is exactly my point. The product already exists and the open source part is still just a lip service.

**Comment by scpcom** [comment]
> > but can not compared with the project(s) discussed here: providing a full firmware image for an existing product.
> 
> And that is exactly my point. The product already exists and the open source part is still just a lip service.

For me a kickstarter project like jetkvm is "just a lip service" by design. They may do it better but at this moment we can just guess.
The base firmware image for NanoKVM was available from the beginning. The server/web sources followed later.
The sources for original maixcam_lib are still missing but they provided enough code to which made it possible to reproduce a fully functional replacement.

**Comment by scpcom** [comment]
The source code for sipeed kvm_system was added here now:
https://github.com/sipeed/NanoKVM/commit/1db304a5f36e07974ee93bce2758779d5f66fdc5

It is build with MaixCDK, if you use my repository you can enable it with:
`./build-nanokvm.sh --maixcdk`
And you can combine it with tailsale if needed:
`./build-nanokvm.sh --maixcdk --tailscale`

**Comment by oliv3r** [comment]
@scpcom amazing work, really glad someone is picking up here.

You mentioned earlier:
> If the build is done (or you download them from the Releases page)

But this did not accompany a link ;)

Anyway, are you hosting/releasing any images for the nanokvm? I'd like to built it, but the 40+gb disk space requirement makes it a bit hard at the moment ...

**Comment by scpcom** [comment]
> You mentioned earlier:
> 
> > If the build is done (or you download them from the Releases page)
> 
> But this did not accompany a link ;)

I only mentioned it for the toolchain, which can be found here:
https://github.com/scpcom/riscv-gnu-toolchain/releases/tag/riscv64-gcc-thead_20230307-10.2.0
Currently I do not plan to release any images. My focus is still on the source code and to make it as easy as possible to build your own image from it.

**Comment by scpcom** [comment]
Since I added a github workflow, the build process is fully transparent now and you can download the resulting ready-to-use images here:
https://github.com/scpcom/LicheeSG-Nano-Build/releases

**Comment by howels** [comment]
> Since I added a new github workflow (inspired by sophgo-sg200x-debian), the build process is fully transparent now and you can download the resulting ready-to-use stable images here: https://github.com/scpcom/LicheeSG-Nano-Build/releases
> 
> I also started experimenting with Ubuntu (NanoKVM app and HDMI/camera input is already working, but keyboard/mouse control with USB gadget does not work): https://github.com/scpcom/sophgo-sg200x-debian/releases

Do these work with the NanoKVM PCIe model, i.e. does your release include the updates in the official 1.3.0 release?

**Comment by scpcom** [comment]
It contains
- all updates from the NanoKVM repository here (app version 2.1.5 + Download button for ISO images)
- my open source version of the maixam_lib (does work with any device that has the same chip, no difference for PCIe-Slot variant)
- almost all other hardware specific things are based on the LicheeRV-Nano-Build repsitory and the chip vendors SDK

The settings of the PCIe version are done by the init.d scripts (/etc/init.d/S15kvmhwd and more) which are also included.
I do not think the 1.3.0 image itself does contain any extra content.
On the Debian/Ubuntu version I did not implement the WiFi setup yet, all other should work.

I do have a NanoKVM Lite and NanoKVM Full (Cube) for testing.
The NanoKVM PCIe is already shipping to me.

**Comment by howels** [comment]
> It contains
> 
>     * all updates from the NanoKVM repository here (app version 2.1.5 + Download button for ISO images)
> 
>     * my open source version of the maixam_lib (does work with any device that has the same chip, no difference for PCIe-Slot variant)
> 
>     * almost all other hardware specific things are based on the LicheeRV-Nano-Build repsitory and the chip vendors SDK
> 
> 
> The settings of the PCIe version are done by the init.d scripts (/etc/init.d/S15kvmhwd and more) which are also included. I do not think the 1.3.0 image itself does contain any extra content. On the Debian/Ubuntu version I did not implement the WiFi setup yet, all other should work.
> 
> I do have a NanoKVM Lite and NanoKVM Full (Cube) for testing. The NanoKVM PCIe is already shipping to me.

Tried your latest build, initially I thought HDMI capture wasn't working on the nanokvm-pcie but I swapped from H264 to MJPEG and it sprung into life. On H264 I get the test-pattern of coloured stripes.

**Comment by scpcom** [comment]
> Tried your latest build, initially I thought HDMI capture wasn't working on the nanokvm-pcie but I swapped from H264 to MJPEG and it sprung into life. On H264 I initially get the test-pattern of coloured stripes then when swapping back from MJPEG the site is stuck at "loading" with spinning dots.

On NanoKVM Cube I can switch from H264 to MJPEG and vice versa without a problem. The only relevant difference between H264 and MJEPG is: MJPEG needs much more CPU on the client hardware (web browser) side.
The test-pattern is shown if no input is detected.
This also happens if the target system sends the "energy saving"  signal to HDMI. In this case you can press a key or move the mouse over the test-pattern and the HDMI input will show up again.
But maybe the H264 needs more power and NanoKVM gets not enough via PCIe? I will look at it if the hardware arrives.

**Comment by howels** [comment]
> > Tried your latest build, initially I thought HDMI capture wasn't working on the nanokvm-pcie but I swapped from H264 to MJPEG and it sprung into life. On H264 I initially get the test-pattern of coloured stripes then when swapping back from MJPEG the site is stuck at "loading" with spinning dots.
> 
> On NanoKVM Cube I can switch from H264 to MJPEG and vice versa without a problem. The only relevant difference between H264 and MJEPG is: MJPEG needs much more CPU on the client hardware (web browser) side. The test-pattern is shown if no input is detected. This also happens if the target system sends the "energy saving" signal to HDMI. In this case you can press a key or move the mouse over the test-pattern and the HDMI input will show up again. But maybe the H264 needs more power and NanoKVM gets not enough via PCIe? I will look at it if the hardware arrives.

Thanks for getting a device to check, this device is powered via PoE as I haven't put it into a chassis yet. Don't think power is an issue as the PoE readout was well under 10W of draw, I think it was only using 3 or 4W.

**Comment by wj-xiao** [comment]
All the code is now open-source. And this issue is closed.

We'll be adding more documentation and build scripts in the near future. 

For any other matters, please feel free to open a new issue for discussion.

**Comment by DeveeiveD** [comment]
Build scripts when?

**Comment by scpcom** [comment]
The official build instructions and scripts for NanoKVM server are here:
https://github.com/sipeed/NanoKVM/tree/main/support/sg2002
https://github.com/sipeed/NanoKVM/tree/main/server

The official build script for NanoKVM firmware image is here:
https://github.com/sipeed/LicheeRV-Nano-Build/tree/NanoKVM/kvm

**Comment by JohnTrabusca** [comment]
@scpcom any idea where I can get this file mentioned herehttps://github.com/sipeed/LicheeRV-Nano-Build/blob/98d17d2c04f9f6efc393c0e6a93b1a0f126b084b/kvm/NanoKVM_img.sh#L24C8-L24C18 ?

I'm using their build script but that part in specific fails as the files are missing.

**Comment by scpcom** [comment]
I do not use this script in my build (see above).
Maybe it works if you replace "kvm/init.d" with "kvmapp/system/init.d"
